client/allocrunner/consul_hook.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package allocrunner
+
+import (
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/consul"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/client/widmgr"
+	"github.com/hashicorp/nomad/nomad/structs"
+	structsc "github.com/hashicorp/nomad/nomad/structs/config"
+)
+
+const (
+	// consulServicesAuthMethodName is the JWT auth method name that has to be
+	// configured in Consul in order to authenticate Nomad services.
+	consulServicesAuthMethodName = "nomad-workloads"
+
+	// consulTemplatesAuthMethodName the JWT auth method name that has to be
+	// configured in Consul in order to authenticate Nomad templates.
+	consulTemplatesAuthMethodName = "nomad-templates"
+)
+
+type consulHook struct {
+	alloc         *structs.Allocation
+	allocdir      *allocdir.AllocDir
+	widmgr        widmgr.IdentityManager
+	consulConfigs map[string]*structsc.ConsulConfig
+	hookResources *cstructs.AllocHookResources
+
+	logger log.Logger
+}
+
+func newConsulHook(logger log.Logger, alloc *structs.Allocation,
+	allocdir *allocdir.AllocDir,
+	widmgr widmgr.IdentityManager,
+	consulConfigs map[string]*structsc.ConsulConfig,
+	hookResources *cstructs.AllocHookResources,
+) *consulHook {
+	h := &consulHook{
+		alloc:         alloc,
+		allocdir:      allocdir,
+		widmgr:        widmgr,
+		consulConfigs: consulConfigs,
+		hookResources: hookResources,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*consulHook) Name() string {
+	return "consul"
+}
+
+func (h *consulHook) Prerun() error {
+	job := h.alloc.Job
+
+	if job == nil {
+		// this is always a programming error
+		err := fmt.Errorf("alloc %v does not have a job", h.alloc.Name)
+		h.logger.Error(err.Error())
+		return err
+	}
+
+	mErr := multierror.Error{}
+
+	// tokens are a map of Consul cluster to service/identity name to Consul
+	// ACL token
+	tokens := map[string]map[string]string{}
+
+	// get consul clients
+	clients, err := getConsulClients(h.consulConfigs, h.logger)
+	if err != nil {
+		return err
+	}
+
+	for _, tg := range job.TaskGroups {
+		for _, service := range tg.Services {
+			t, err := h.prepareConsulTokens(service, clients)
+			if err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+				continue
+			}
+			tokens[service.Cluster] = t
```

Comment:
If we use `service.Cluster` as the key, then if we have multiple services in the same cluster in the same group/task we'd overwrite the token. And each token has its own policies generated on the Consul side, so that won't work. I think we need to key these by something unique to the service in this allocation. (Ex. the group + task + service names or something like that.)